openslides_backend/action/mediafile/move.py

```diff
@@ -0,0 +1,87 @@
+class MediafileMoveAction(UpdateAction):
+    """
+    Action to move mediafiles.
+    """
+
+    model = Mediafile()
+    schema = DefaultSchema(Mediafile()).get_default_schema(
+        title="Mediafile new parent schema",
+        description="An object containing an array of mediafile ids and the new parent id the items should be moved to.",
+        required_properties=["parent_id", "meeting_id"],
+        additional_required_fields={
+            "ids": {
+                "description": "An array of agenda item ids where the items should be assigned to the new parent id.",
+                **id_list_schema,
+            }
+        },
+        single_item=True,
+    )
+
+    def check_is_directory(self, id_: int) -> None:
+        item = self.datastore.get(
+            FullQualifiedId(self.model.collection, id_), ["is_directory"]
+        )
+        if not item.get("is_directory"):
+            raise ActionException("New parent is not a directory.")
+
+    def get_updated_instances(self, payload: ActionPayload) -> ActionPayload:
+        # Payload is an iterable with exactly one item
+        instance = next(iter(payload))
+        yield from self.prepare_assign_data(
+            parent_id=instance["parent_id"],
+            ids=instance["ids"],
+            meeting_id=instance["meeting_id"],
+        )
+
+    def prepare_assign_data(
+        self, parent_id: Optional[int], ids: List[int], meeting_id: int
+    ) -> ActionPayload:
+        filter = FilterOperator("meeting_id", "=", meeting_id)
+        db_instances = self.datastore.filter(
+            collection=self.model.collection,
+            filter=filter,
+            mapped_fields=["id"],
+            lock_result=True,
+        )
+
+        if parent_id is None:
+            for id_ in ids:
+                if id_ not in db_instances:
+                    raise ActionException(f"Id {id_} not in db_instances.")
+                yield {"id": id_, "parent_id": None}
+        else:
+            self.check_is_directory(parent_id)
+
+            # Calculate the ancesters of parent
+            ancesters = [parent_id]
+            grandparent = self.datastore.get(
+                FullQualifiedId(self.model.collection, parent_id), ["parent_id"]
+            )
+            while grandparent.get("parent_id") is not None:
+                gp_parent_id = grandparent["parent_id"]
+                ancesters.append(gp_parent_id)
+                grandparent = self.datastore.get(
+                    FullQualifiedId(self.model.collection, gp_parent_id), ["parent_id"]
+                )
+            for id_ in ids:
+                if id_ in ancesters:
+                    raise ActionException(
+                        f"Moving item {id_} to one of its children is not possible."
+                    )
+                if id_ not in db_instances:
+                    raise ActionException(f"Id {id_} not in db_instances.")
+                yield {"id": id_, "parent_id": parent_id}
```

Comment:
You may also have to update the `inherited_access_group_ids` as well as `is_public`, depending ob the `access_group_ids` from children and parent. Also add tests for this.

Reply:
Calculated is_public and inherited_access_group_id for item and children, used the mediafile update code for this.

Reply:
Also add and update tests.